js/ui/status/keyboard.js

```diff
@@ -480,6 +484,52 @@ const InputSourceManager = new Lang.Class({
             this._updateMruSettings();
     },
 
+    get passwordModeEnabled() {
+        return this._passwordModeEnabled;
+    },
+
+    set passwordModeEnabled(enable) {
+        if (this._passwordModeEnabled === enable)
+            return;
+
+        this._passwordModeEnabled = enable;
+        this._syncPasswordMode();
+    },
+
+    _syncPasswordMode: function() {
+        let source;
+
+        if (this._passwordModeEnabled) {
+            // do not switch the keyboard layout if the current one is a latin one
+            if (this._keyboardManager.isLatinLayout(this._currentSource.id))
+                return;
+
+            // look for latin keyboard layout to switch to
+            for (let i = 0; i < this._mruSources.length; ++i) {
+                let src = this._mruSources[i];
+                if (this._keyboardManager.isLatinLayout(src.id)) {
+                    source = src;
+                    break;
+                }
+            }
+            if (!source) {
+                // fallback to an 'en' keyboard layout
+                this._inputSourcesChanged(true);
+                source = this._fallbackInputSource;
+            }
+            this._lastSource = this._currentSource;
+            this._activateInputSource(source, false);
+        } else {
+            if (!this._lastSource)
+                return;
+            source = this._lastSource;
+            this._activateInputSource(source, false);
+            this._lastSource = null;
+            this._fallbackInputSource = null;
```

Comment:
I think it would be cleaner if this happened inside `this._inputSourcesChanged()` when `needsFallbackSource == false`

Reply:
Done.